distrho/src/DistrhoPluginVST.cpp

```diff
@@ -280,8 +280,28 @@ class UIVst
 
     void sendNote(const uint8_t channel, const uint8_t note, const uint8_t velocity)
     {
-# if 0 //DISTRHO_PLUGIN_WANT_MIDI_INPUT
-        // TODO
+# if DISTRHO_PLUGIN_WANT_MIDI_INPUT
+        if (channel > 0xF)
+            return;
+
+        VstMidiEvent midiEvent;
+        memset(&midiEvent, 0, sizeof(VstMidiEvent));
+
+        midiEvent.type = kVstMidiType;
+        midiEvent.byteSize = sizeof(VstMidiEvent);
+
+        midiEvent.midiData[0] = channel;
+        midiEvent.midiData[0] += (velocity != 0) ? 0x90 : 0x80;
+        midiEvent.midiData[1] = note;
+        midiEvent.midiData[2] = velocity;
+
+        VstEvents event;
+        memset(&event, 0, sizeof(VstEvents));
+
+        event.numEvents = 1;
+        event.events[0] = (VstEvent*)&midiEvent;
+
+        fEffect->dispatcher(fEffect, effProcessEvents, 0, 0, &event, 0.0f);
```

Comment:
This is nasty... effProcessEvents is supposed to be called by the plugin and only during processing...
Did you try this with carla? very likely carla will see it being called from the wrong thread.

It needs a different approach I believe.

Reply:
Carla seems happy with it. 

Anyway, what would be a good approach?

Reply:
a ring buffer with a read-try-lock on the process/RT side of things.

dpf does not have such thing.
I dont think there will be a lot of events from the UI ever, as RT side can get them pretty quickly.
so perhaps just a static array of 16 events, with a mutex that does tryLock() on vst process would work fine.